src/main/java/ui/components/pickers/LabelPickerUILogic.java

```diff
@@ -59,58 +71,146 @@ private void populatePanes() {
+        }
 
-                if (hasHighlightedLabel()) {
-                    addRemovePossibleLabel(getHighlightedLabelName().get().getActualName());
-                } else {
-                    addRemovePossibleLabel("");
-                }
-                populatePanes();
-            }
+        updateVariables(text);
+    }
+
+    /**
+     * This method will either toggle the highlighted label, or update both the bottom labels and the possible label
+     * in the top pane, depending on the last character of text (added character)
+     */
+    private void addCharHandler(String text) {
+        if (isLastCharSpace(text)) {
+            // space-toggle
+            spaceToggle();
+        } else {
+            queryHandler(getLastWord(text));
         }
     }
 
+    private void spaceToggle() {
+        if (!bottomLabels.isEmpty() && hasHighlightedLabel()) {
+            stackOfSpecifiedLabels.push(getHighlightedLabelName().get().getActualName());
+            stackOfRemovedExclusiveLabels.push(toggleHighlightedLabel());
+        } else {
+            stackOfSpecifiedLabels.push(null);
+            stackOfRemovedExclusiveLabels.push(null);
+        }
+    }
+
+    private void queryHandler(String lastWord) {
+        updateBottomLabelsWithRawQuery(lastWord);
+        updatePossibleLabel();
+    }
+
+    private void removeCharHandler(String text) {
+        if (isWordRemoved(text)) {
+            undoPreviousSpaceToggle();
+        }
+
+        if (isLastCharSpace(text)) {
+            removePossibleLabel();
+            restoreBottomLabels();
+        } else {
+            queryHandler(getLastWord(text));
+        }
+    }
+
+    private void restoreBottomLabels() {
+        updateBottomLabels("");
+    }
+
+    private void undoPreviousSpaceToggle() {
+        toggleRecentlySpecifiedLabel();
+        restoreRecentlyRemovedLabels();
+    }
+
+    /**
+     * Untoggles last label that was specified
+     * This may do nothing if the last label added was null (i.e. added without any highlighted label)
+     */
+    private void toggleRecentlySpecifiedLabel() {
+        String lastSpecifiedLabel = stackOfSpecifiedLabels.pop();
+        if (lastSpecifiedLabel != null) {
+            // if the last label is mapped to an actual label
+            updateTopLabels(lastSpecifiedLabel, !resultList.get(lastSpecifiedLabel));
+        }
+    }
+
+    /**
+     * Restores the most recent set of removed label(s) due to exclusive grouping
+     */
+    private void restoreRecentlyRemovedLabels() {
+        ArrayList<String> listOfRemovedLabels = stackOfRemovedExclusiveLabels.pop();
+        if (listOfRemovedLabels != null) {
+            listOfRemovedLabels.stream()
+                    .forEach(labelName -> updateTopLabels(labelName, true));
+        }
+    }
+
+    private boolean isEmpty(String text) {
+        return text.length() == 0;
+    }
+
+    private boolean isLongerTextFieldThanPrevious(String text) {
```

Comment:
I think the purpose of this method is to check if a char was added. Checking the length is the mechanism for doing that. The method name should reflect the intent (WHAT it does), not mechanism. You can name it `isCharAdded(String previousText, String newText)`. By feeding both parameters, you remove a hidden dependency between the method and an instance variable.